core/src/main/codegen/templates/Parser.jj

```diff
@@ -5050,6 +5050,86 @@ SqlIntervalQualifier TimeUnitOrName() : {
     }
 }
 
+  SqlIntervalQualifier TimeUnitOrNameForDateTrunc() : {
+    final TimeUnit timeUnit;
+    final SqlIntervalQualifier weekdayIntervalQualifier;
+    final SqlIdentifier unitName;
+}
+{
+    LOOKAHEAD(1)
+    timeUnit = TimeUnitForDateTrunc() {
+        return new SqlIntervalQualifier(timeUnit, null, getPos());
+    }
+|
+    weekdayIntervalQualifier = WeekWeekdaySqlIntervalQualifier() {
+        return weekdayIntervalQualifier;
+    }
+|
+    unitName = DateTruncIdentifiers() {
+        return new SqlIntervalQualifier(unitName.getSimple(),
+          unitName.getParserPosition());
+    }
+}
+
+/** Parses time unit for the DATE_TRUNC function.
+* This is a subset of time units appropriate for DATE data type.
+*/
+TimeUnit TimeUnitForDateTrunc() : {}
+{
+    <DAY> { return TimeUnit.DAY; }
+|   <MONTH> { return TimeUnit.MONTH; }
+|   <QUARTER> { return TimeUnit.QUARTER; }
+|   <YEAR> { return TimeUnit.YEAR; }
+}
+
+SqlIntervalQualifier WeekWeekdaySqlIntervalQualifier() : {
+    final String weekdayName;
+}
+{
+    weekdayName =  WeekWeekday() {
+        return new SqlIntervalQualifier(weekdayName, getPos());
+    }
+}
+String WeekWeekday() :
+{
+}
+{
+  LOOKAHEAD(4)
+   <WEEK> <LPAREN> <SUNDAY> <RPAREN> { return "WEEK_SUNDAY"; }
+|
+   LOOKAHEAD(4)
+   <WEEK> <LPAREN> <MONDAY> <RPAREN> { return "WEEK_MONDAY"; }
+|
+   LOOKAHEAD(4)
+   <WEEK> <LPAREN> <TUESDAY> <RPAREN> { return "WEEK_TUESDAY"; }
+|
+    LOOKAHEAD(4)
+    <WEEK> <LPAREN> <WEDNESDAY> <RPAREN> { return "WEEK_WEDNESDAY"; }
+|
+    LOOKAHEAD(4)
+    <WEEK> <LPAREN> <THURSDAY> <RPAREN> { return "WEEK_THURSDAY"; }
+|
+    LOOKAHEAD(4)
+    <WEEK> <LPAREN> <FRIDAY> <RPAREN> { return "WEEK_FRIDAY"; }
+|
+    LOOKAHEAD(4)
+    <WEEK> <LPAREN> <SATURDAY> <RPAREN> { return "WEEK_SATURDAY"; }
+|
+    LOOKAHEAD(1)
+    <WEEK> { return "WEEK"; }
+}
```

Comment:
High lookahead values are expensive, and LOOKAHEAD(4) might be overkill. I would write something like

```
<WEEK>
(
  <LPAREN> w = weekDay() <RPAREN> { return weekDays.get(w); }
|
  /* empty */ { return "WEEK"; }
)
```
and now the parser doesn't have to go so deep before it makes a decision.